aw-datastore/src/datastore.rs

```diff
@@ -581,4 +598,84 @@ impl DatastoreInstance {
 
         return Ok(count);
     }
+
+    pub fn insert_value(&self, conn: &Connection, key: &str, data: &str)
+        -> Result<(), DatastoreError> {
+
+        let mut stmt = match conn.prepare("
+                INSERT OR REPLACE INTO key_value(key, value)
+                VALUES (?1, ?2)") {
+            Ok(stmt) => stmt,
+            Err(err) => return Err(DatastoreError::InternalError(
+                format!("Failed to prepare insert_value SQL statement: {}", err)
+            ))
+        };
+        stmt.execute(&[key, data]).expect(
+            &format!("Failed to insert key-value pair: {}", key)
+        );
+        return Ok(())
+    }
+
+    pub fn delete_value(&self, conn: &Connection, key: &str) -> Result<(), DatastoreError>{
+        conn.execute("DELETE FROM key_value WHERE key = ?1", &[key])
+            .expect("Error deleting value from database");
+        return Ok(())
+    }
+
+    pub fn get_value(&self, conn: &Connection, key: &str) -> Result<String, DatastoreError>{
+        let mut stmt = match conn.prepare("
+                SELECT * FROM key_value WHERE KEY = ?1") {
+            Ok(stmt) => stmt,
+            Err(err) => return Err(DatastoreError::InternalError(
+                format!("Failed to prepare get_value SQL statement: {}", err)
+            ))
+        };
+
+        return match stmt.query_row(&[key], |row|{
+            row.get(1)
+        }) {
+            Ok(result)  => Ok(result),
+            Err(err) => match err {
+                rusqlite::Error::QueryReturnedNoRows => Err(DatastoreError::NoSuchValue),
+                _ => Err(DatastoreError::InternalError(
+                              format!("Get value query failed for key {}", key))
+                )
+            }
+        }
+    }
+
+    pub fn get_values_starting(&self, conn: &Connection, pattern: &str) 
+        -> Result<HashMap<String, String>, DatastoreError>
+    {
+        let mut stmt = match conn.prepare("SELECT * FROM key_value WHERE key LIKE '.%' ") {
+            Ok(stmt) => stmt,
+            Err(err) => return Err(DatastoreError::InternalError(
+                format!("Failed to prepare get_value SQL statement: {}", err)
+            ))
+        };
+
+        let mut output = HashMap::<String, String>::new();
+        // Unwrap to String tuple or panic on SQL row if types are invalid.
+        // should never happen with a properly initialized table
+        let result = stmt.query_map(&[pattern], |row| Ok(
+            (row.get::<usize, String>(0).expect("Invalid sql column types"), 
```

Comment:
What's up with the usize here? Isn't it supposed to just get the strings key and value? What is the usize for?

Reply:
I couldn't get row.get to work with just one type parameter, it also wanted a type param for the index which is an usize. There should be a way to give in now type param and force it to a string, but I didn't figure out one when writing this.

Reply:
Fixed in 8277c1b